File: agenkit-cpp/src/skills/skill_enabled_agent.cpp
/**
 * @file skill_enabled_agent.cpp
 * @brief SkillEnabledAgent implementation.
 */

#include "agenkit/skills/skill_enabled_agent.hpp"

#include <algorithm>
#include <utility>

namespace agenkit {
namespace skills {

SkillEnabledAgent::SkillEnabledAgent(std::shared_ptr<core::Agent> agent,
                                     std::shared_ptr<SkillRegistry> registry,
                                     std::size_t max_active_skills,
                                     bool auto_discover)
    : agent_(std::move(agent)),
      registry_(std::move(registry)),
      max_active_skills_(max_active_skills) {
    if (auto_discover && registry_) {
        registry_->discover_skills();
    }
}

std::string SkillEnabledAgent::name() const {
    return agent_->name();
}

std::vector<std::string> SkillEnabledAgent::capabilities() const {
    std::vector<std::string> caps = agent_->capabilities();
    if (std::find(caps.begin(), caps.end(), "skill_injection") == caps.end()) {
        caps.push_back("skill_injection");
    }
    return caps;
}

std::future<core::Result<core::Message, core::AgentError>>
SkillEnabledAgent::process(core::Message message) {
    const std::string query = message.content_as_str();

    std::vector<AgentSkill> relevant;
    if (registry_) {
        relevant = registry_->find_relevant_skills(query, max_active_skills_);
    }

    if (relevant.empty()) {
        return agent_->process(std::move(message));
    }

    // Build the <available_skills> block from the relevant skills.
    std::string skill_blocks;
    for (std::size_t i = 0; i < relevant.size(); ++i) {
        if (i > 0) {
            skill_blocks += "\n\n";
        }
        skill_blocks += relevant[i].to_prompt();
    }

    const std::string augmented_content =
        "<available_skills>\n" + skill_blocks + "\n</available_skills>\n\n" + query;

    // Preserve existing metadata, then record the active skill names.
    nlohmann::json metadata = message.metadata();
    nlohmann::json active = nlohmann::json::array();
    for (const auto& skill : relevant) {
        active.push_back(skill.name);
    }

    core::Message enhanced(message.role(), nlohmann::json(augmented_content));
    if (metadata.is_object()) {
        for (auto it = metadata.begin(); it != metadata.end(); ++it) {
            enhanced.with_metadata(it.key(), it.value());
        }
    }
    enhanced.with_metadata("active_skills", active);

    return agent_->process(std::move(enhanced));
}

} // namespace skills
} // namespace agenkit

File: agenkit-cs/src/Agenkit/Skills/SkillEnabledAgent.cs
using Agenkit.Core;

namespace Agenkit.Skills;

/// <summary>
/// Agent wrapper that automatically injects relevant skill instructions.
///
/// Before delegating to the wrapped agent, this wrapper queries the registry
/// for skills relevant to the incoming message and prepends their instructions
/// inside an <c>&lt;available_skills&gt;</c> block. The augmented message's
/// metadata contains <c>active_skills</c> listing the injected skill names.
/// </summary>
public sealed class SkillEnabledAgent : IAgent
{
    private readonly IAgent _agent;
    private readonly SkillRegistry _registry;
    private readonly int _maxActiveSkills;

    /// <summary>Creates a skill-enabled agent wrapping <paramref name="agent"/>.</summary>
    /// <param name="agent">Base agent to delegate processing to.</param>
    /// <param name="registry">Registry used to look up relevant skills.</param>
    /// <param name="maxActiveSkills">Maximum number of skills to inject (default 3).</param>
    /// <param name="autoDiscover">Whether to call <see cref="SkillRegistry.DiscoverSkills"/> at construction (default true).</param>
    public SkillEnabledAgent(
        IAgent agent,
        SkillRegistry registry,
        int maxActiveSkills = 3,
        bool autoDiscover = true)
    {
        _agent = agent;
        _registry = registry;
        _maxActiveSkills = maxActiveSkills;
        if (autoDiscover)
            _registry.DiscoverSkills();
    }

    /// <inheritdoc />
    public string Name => _agent.Name;

    /// <inheritdoc />
    public IReadOnlyList<string> Capabilities
    {
        get
        {
            var caps = new List<string>(_agent.Capabilities);
            if (!caps.Contains("skill_injection"))
                caps.Add("skill_injection");
            return caps;
        }
    }

    /// <inheritdoc />
    public IntrospectionResult Introspect() => _agent.Introspect();

    /// <inheritdoc />
    public Task<Message> ProcessAsync(Message message, CancellationToken ct = default)
    {
        var query = message.ContentString();
        var relevant = _registry.FindRelevantSkills(query, _maxActiveSkills);

        if (relevant.Count == 0)
            return _agent.ProcessAsync(message, ct);

        var skillBlocks = string.Join("\n\n", relevant.Select(s => s.ToPrompt()));
        var augmentedContent = $"<available_skills>\n{skillBlocks}\n</available_skills>\n\n{query}";

        var metadata = message.Metadata is not null
            ? new Dictionary<string, object>(message.Metadata)
            : new Dictionary<string, object>();
        metadata["active_skills"] = relevant.Select(s => s.Name).ToList();

        var enhanced = message with { Content = augmentedContent, Metadata = metadata };
        return _agent.ProcessAsync(enhanced, ct);
    }
}
